gpytorch/models/exact_prediction_strategies.py

```diff
@@ -365,28 +364,88 @@ def _exact_predictive_covar_inv_quad_form_root(self, precomputed_cache, test_tra
         return res
 
     def get_fantasy_strategy(self, inputs, targets, full_inputs, full_targets, full_output, **kwargs):
-        raise NotImplementedError(
-            "Fantasy observation updates not yet supported for models using InterpolatedLazyTensors"
+        r"""
+        Implements the fantasy strategy described in https://arxiv.org/abs/2103.01454.
+        """
+        full_mean, full_covar = full_output.mean, full_output.lazy_covariance_matrix
+
+        batch_shape = full_inputs[0].shape[:-2]
+
+        full_mean = full_mean.view(*batch_shape, -1)
+        num_train = self.num_train
+
+        # Evaluate fant x train and fant x fant covariance matrices, leave train x train unevaluated.
+        fant_fant_covar = full_covar[..., num_train:, num_train:]
+        # fant_mean = full_mean[..., num_train:]
+
+        fant_wmat = self.prepare_dense_wmat(fant_fant_covar)
+
+        fant_likelihood = self.likelihood.get_fantasy_likelihood(**kwargs)
+        fant_noise = fant_likelihood.noise_covar(inputs)
+        fant_root_vector = fant_noise.sqrt_inv_matmul(fant_wmat.transpose(-1, -2)).transpose(-1, -2)
+
+        new_wmat = self.interp_inner_prod.add_low_rank(fant_root_vector)
+        new_interp_response_cache = self.interp_response_cache + new_wmat.matmul(fant_noise.inv_matmul(targets))
+
+        # Create new DefaultPredictionStrategy object
+        fant_strat = self.__class__(
+            train_inputs=full_inputs,
+            train_prior_dist=self.train_prior_dist.__class__(full_mean, full_covar),
+            train_labels=full_targets,
+            likelihood=fant_likelihood,
         )
+        add_to_cache(fant_strat, "interp_inner_prod", new_wmat)
+        add_to_cache(fant_strat, "interp_response_cache", new_interp_response_cache)
+
+    def prepare_dense_wmat(self, covar=None):
```

Comment:
Just so I'm clear here - this method is only called with `fant_fant_covar`, so we can expect these matrices to be small. Is that correct?

Reply:
It also has to be done on the training distribution so it could be a dense matrix of shape `m x n`.